File: go/hashmap/349_intersection.go
package hashmap

// intersection 给定两个数组 nums1 和 nums2 ，返回 它们的交集 。
// 输出结果中的每个元素一定是 唯一 的。我们可以 不考虑输出结果的顺序 。
func intersection(nums1 []int, nums2 []int) []int {
	// hashmap 记录某个数组中的元素，然后和另一个数组进行比较
	hmap := make(map[int]struct{})

	for _, v := range nums1 {
		hmap[v] = struct{}{}
	}

	hmapResult := make(map[int]struct{})
	for _, v := range nums2 {
		if _, ok := hmap[v]; ok {
			if _, ok := hmapResult[v]; !ok {
				hmapResult[v] = struct{}{}
			}
		}
	}

	result := make([]int, 0)
	for k := range hmapResult {
		result = append(result, k)
	}
	return result
}

File: rust/hashmap/349_intersection.rs


pub fn intersection(nums1: Vec<i32>, nums2: Vec<i32>) -> Vec<i32> {
    // 集合
    use std::collections::HashSet;
    let mut set = HashSet::new();

    for &v in &nums1 {
        set.insert(v);
    }

    let mut result = vec![];
    let mut set2 = HashSet::new();

    for &v in nums2.iter() {
        if set.contains(&v) && !set2.contains(&v) {
            result.push(v);
        }
        set2.insert(v);
    }
    result
}

pub fn intersection2(nums1: Vec<i32>, nums2: Vec<i32>) -> Vec<i32> {
    // 集合
    use std::collections::HashSet;
    // let set: HashSet<i32> = nums1.iter().cloned().collect();
    let set: HashSet<i32> = nums1.into_iter().collect();

    let mut result_set = HashSet::new();

    for &v in nums2.iter() {
        if set.contains(&v) {
            result_set.insert(v);
        }
    }

    result_set.into_iter().collect()
}

pub fn intersection3(nums1: Vec<i32>, nums2: Vec<i32>) -> Vec<i32> {
 
    use std::collections::HashSet;
    nums1.into_iter().collect::<HashSet<i32>>()
    .intersection(&nums2.into_iter().collect::<HashSet<i32>>())
    .map(|&x|x)
    .collect()

}